proxy/src/registry/transaction.rs

```diff
@@ -0,0 +1,338 @@
+//! Abstractions and types to handle, persist and interact with transactions.
+
+use async_trait::async_trait;
+use hex::ToHex;
+use kv::Codec as _;
+use serde::{Deserialize, Serialize};
+use std::time::SystemTime;
+
+use radicle_registry_client as protocol;
+
+use crate::error;
+use crate::registry;
+
+/// A container to dissiminate and apply operations on the [`Registry`].
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Transaction {
+    /// Unique identifier, in actuality the Hash of the transaction. This handle should be used by
+    /// the API consumer to query state changes of a transaction.
+    pub id: registry::Hash,
+    /// List of operations to be applied to the Registry. Currently limited to exactly one. We use
+    /// a Vec here to accommodate future extensibility.
+    pub messages: Vec<Message>,
+    /// Current state of the transaction.
+    pub state: State,
+    /// Creation time.
+    pub timestamp: SystemTime,
+}
+
+/// Possible messages a [`Transaction`] can carry.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase", tag = "type")]
+pub enum Message {
+    /// Issue a new org registration.
+    OrgRegistration {
+        /// The [`registry::Org`] id.
+        id: registry::Id,
+    },
+
+    /// Issue an org unregistration with a given id.
+    OrgUnregistration {
+        /// The [`registry::Org`] id.
+        id: registry::Id,
+    },
+
+    /// Issue a new project registration with a given name under a given org.
+    ProjectRegistration {
+        /// Actual project name, unique for org.
+        project_name: registry::ProjectName,
+        /// The Org in which to register the project.
+        org_id: registry::Id,
+    },
+
+    /// Issue a user registration for a given handle storing the corresponding identity id.
+    UserRegistration {
+        /// Globally unique user handle.
+        handle: registry::Id,
+        /// Identity id originated from librad.
+        id: Option<String>,
+    },
+}
+
+/// Possible states a [`Transaction`] can have. Useful to reason about the lifecycle and
+/// whereabouts of a given [`Transaction`].
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase", tag = "type")]
+pub enum State {
+    /// [`Transaction`] has been applied to a block, carries the hash of the block.
+    Applied {
+        /// The hash of the block the transaction has been applied to.
+        block: registry::Hash,
+    },
+}
+
+/// Behaviour to manage and persist observed [`Transaction`].
+#[async_trait]
+pub trait Cache: Send + Sync {
+    /// Caches a transaction locally in the Registry.
+    async fn cache_transaction(&mut self, tx: Transaction) -> Result<(), error::Error>;
+
+    /// Returns all cached transactions.
+    ///
+    /// # Errors
+    ///
+    /// Will return `Err` if a protocol error occurs.
+    async fn list_transactions(
+        &self,
+        ids: Vec<protocol::TxHash>,
```

Comment:
Does the function need to own the TxHashes? I'd expect this to take a slice normally.